language/libra-vm/src/libra_vm.rs

```diff
@@ -332,115 +354,118 @@ impl LibraVM {
+                    .publishing_option
+                    .is_open()
+                {
+                    txn_data.sender()
+                } else {
+                    account_config::CORE_CODE_ADDRESS
+                };
+                unwrap_or_fail!(self.move_vm.publish_module(
+                    m,
+                    module_address,
+                    &mut data_store,
+                    &mut cost_strategy,
+                ))
+            }
             VerifiedTransactionPayload::Script(s, ty_args, args) => {
-                let ret = cost_strategy
-                    .charge_intrinsic_gas(txn_data.transaction_size())
-                    .and_then(|_| {
-                        self.move_vm.execute_script(
-                            s,
-                            ty_args,
-                            args,
-                            txn_data.sender(),
-                            &mut data_store,
-                            &mut cost_strategy,
-                        )
-                    });
+                unwrap_or_fail!(cost_strategy.charge_intrinsic_gas(txn_data.transaction_size()));
+                unwrap_or_fail!(self.move_vm.execute_script(
+                    s,
+                    ty_args,
+                    args,
+                    txn_data.sender(),
+                    &mut data_store,
+                    &mut cost_strategy,
+                ));
                 let gas_usage = txn_data
                     .max_gas_amount()
                     .sub(cost_strategy.remaining_gas())
                     .get();
                 TXN_EXECUTION_GAS_USAGE.observe(gas_usage as f64);
-                ret
             }
         }
-        .map_err(|err| {
-            failed_gas_left = cost_strategy.remaining_gas();
-            err
-        })
-        .and_then(|_| {
-            failed_gas_left = cost_strategy.remaining_gas();
-            let mut cost_strategy = CostStrategy::system(gas_schedule, failed_gas_left);
-            self.run_success_epilogue(
-                &mut data_store,
-                &mut cost_strategy,
-                txn_data,
-                account_currency_symbol,
-            )
-            .and_then(|_| {
-                get_transaction_output(
-                    &mut data_store,
-                    &cost_strategy,
-                    txn_data,
-                    VMStatus::new(StatusCode::EXECUTED),
-                )
-            })
-        })
-        .unwrap_or_else(|err| {
-            self.failed_transaction_cleanup(
-                err,
-                gas_schedule,
-                failed_gas_left,
-                txn_data,
-                remote_cache,
-                account_currency_symbol,
-            )
-        })
+
+        let mut cost_strategy = CostStrategy::system(gas_schedule, cost_strategy.remaining_gas());
+        unwrap_or_fail!(self.run_success_epilogue(
```

Comment:
do we really need the macros here? just curious...

Reply:
Hmm... Just to clarify, are you questioning the existence of the macro itself, or certain use cases here?

Reply:
I am questioning the usage of the macro altogether yeah :)

Reply:
It's used to emulate a "?" operator with custom behavior. I also had a discussion with @tnowacki as he also ran into this issue. It seems to me this is the most sensible solution that exists in stable rust for now:
-  `?` cannot be overridden  
- `try_blocks` is only available on nightly
- Combinators are hard to understand and cannot handle mutability
  - IMO it's much easier to explicitly sequence the side effects in this case